### PR TITLE
[dashboard] allow editing user information

### LINKED ISSUE #10999: Allow User Profile Updates
Currently, the user information (name and email) is taken from whatever oauth provider you started out with. User can't update them later on. For product-led-sales reasons we want to encourage users to put in their work email and company name.

# Solution

- In addition to `Name` and `Work E-Mail`add a third field `Company` and allow users editing all three.
- [SaaS only] Nudge users once a year that have at least started two workspaces in the past, with a modal in the dashboard to update their profile information (in-modal).

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
This change allows users to update their name, email and company name.
It also nudges users in the workspaces view on the second day (or after) and then once a year.

## Related Issue(s)

Fixes #10999

## How to test
Use the preview env, 
 - create a new user
 - verify you don't get nudged (no modal) on /workspaces
 - set system time to yesterday and reload page. You should now see a modal.
 - either update your profile or simply close the modal (cancel or close).
 - reload page, you should not see the modal anymore (only after one year, which you can verify again by setting your system date)
 - Check the settings page

## Release Notes

```release-note
You can now update your profile information (name, email, company)
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@svenefftinge: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:150]: **issue**: We probably don't need this part in the modal, right? We could simplify the paragraph to resolve this issue.

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:144]: **suggestion**: In case we leave this section here as is, regardless if we simplify it, using the same heading style here is hurting information hierarchy as it conflicts with the modal title. Ideally, dropping this in the modal would be great.

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:120]: **nitpick**: Minor spacing issue with modal action buttons.

```suggestion
                        <button className="ml-2" onClick={saveProfileState}>Update Profile</button>
```

File: components/dashboard/src/settings/Account.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Account.tsx:92]: **suggestion**: There's no need to disable this button. We can safely keep it enabled at all times, right? If so, shall we also change anything else related to `profileState`?

```suggestion
```

File: components/dashboard/src/settings/Account.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Account.tsx:103]: **suggestion**: If we're going with the always enabled primary button, we can safely remove this button altogether.

```suggestion
```

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:145]:
```suggestion
            <p className="text-base text-gray-500 pb-4 max-w-xl">
```

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:152]: **nitpick**: Minor layout change for adding visual balance, based on previous work on defining breakpoints, etc, see relevant comment in https://github.com/gitpod-io/gitpod/issues/8133#issuecomment-1039013543 for more context.

```suggestion
                <div className="dark:bg-gray-800 bg-gitpod-kumquat-light rounded-md p-3 text-gitpod-red text-sm mb-2 max-w-xl">
```

| BEFORE | AFTER |
|-|-|
| Screenshot 2022-06-30 at 6 25 15 PM | Screenshot 2022-06-30 at 6 25 24 PM  |

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:159]: **suggestion**: Ideally, this can safely use the new **[alert component](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/components/Alert.tsx)** while :a: using the message variant and :b: being closable. However, we can leave this out of the scope of this PR for now and create a follow-up issue.

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:79]: **issue(non-blocking)**: This does not safely filter emails with `+` like `user+suffix@company.com`.

File: components/dashboard/src/settings/Account.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/Account.tsx:89]: **nitpick**: Minor spacing shift for better grouping.

```suggestion
                    <div className="flex flex-row mt-8">
```

**Review by gtsiolis — APPROVED** [review]
UX LAGTM (Looks Almost Good To Me)! Thanks, @svenefftinge! 

Cross-posting a relevant issue in https://github.com/gitpod-io/gitpod/issues/3957 in case it can be also closed.

Left some minor comments below. We can open a follow-up issue for anything that can be skipped in this iteration.

**Review by jldec — APPROVED** [review]
LGTM - I think this is a good enhancement. A few comments.

- Users ask to change their avatar almost as often as their name. (followup feature?)

- Offering to change the name and emai during **sign-up** flow is OK and may be a good opportunity to remind new users of our terms and privacy policy.

- I was unable to test whether this modal is triggered when an additional git integration is added through settings/integrations. Generally adding an integration should not result in modification of this information.

Including the hints for how to set env vars to override git email/author/committer is helpful, and I would not remove it.

/hold for more technical engineering review

**Comment by jldec** [comment]
Followup: I managed to test adding an integration using a non-incognito window, and it did not prompt for profile changes - so all good. 

(The error I saw earlier when trying to add an integration was using a preview environment from an incognito window.)

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (svenefftinge) [reply components/dashboard/src/settings/ProfileInformation.tsx:153]: It's a little tough for the eyes, but I agree with @jldec that this information is valuable and we should keep it.

Comment (svenefftinge) [reply components/dashboard/src/settings/ProfileInformation.tsx]: I think we should keep it. It's actually great for users to get reminded with what name and email they are going to commit when using Gitpod.

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (svenefftinge) [reply components/dashboard/src/settings/ProfileInformation.tsx:155]: applied suggestion as well as moved to <Alert />

Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx]: 💯

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:125]: **nitpick**: Spacing fix according to https://github.com/gitpod-io/gitpod/pull/11023#discussion_r911144123, following the spacing pattern used across the product to separate related elements. `ml-4` can be used to separate unrelated elements..

```suggestion
                        <button className="ml-2" onClick={saveProfileState}>
```

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx]: DEAL—Let's keep it for now. 🤝

Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx]: It is indeed somewhat valuable—let's keep it.

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [inline components/dashboard/src/settings/ProfileInformation.tsx:158]: **issue(non-blocking)**: The closable icon is not aligned vertically in the middle but that's out of the scope of this PR. Good news is that now we can fix once and update all closable alert component instances.

**Review by gtsiolis — COMMENTED** [review]
Left one more round of comments, mostly non-blocking.  🥊
Thanks for tackling all these in one step! 🙏

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (svenefftinge) [reply components/dashboard/src/settings/ProfileInformation.tsx:158]: Do you want this to be always centered even when the contents is longer?
Screenshot 2022-07-01 at 12 36 47

Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx:162]: Aligning these to the top would be great. 🔝 

Also, ideally, both the icon on the left, and the closable button on the right use the same bounding box (icon live area + padding) height (`20px`) as the line height of the paragraph text in the middle so that the first line always align with the two icons horizontally, see screenshot below.

alert

Comment (svenefftinge) [reply components/dashboard/src/settings/ProfileInformation.tsx:162]: Ok, the reason the x was a bitz off is that we used a h-3 instead of h-4. I've fixed this now.

Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx:162]: Perfect! One last point that could be part of a follow-up issue or PR is not hide the alert after updating the profile when the user is updating any of the values in the inputs below, just like it's done with the error alert component. The alert should be hidden only when the user explicitly closes (dismiss) the alert or moves to another page away from **/account**. Other than that, this looks ready-to-merge! 🛹 

https://user-images.githubusercontent.com/120486/176913316-17d17d34-f077-4158-b324-d94fb6fae789.mov

**Review by gtsiolis — COMMENTED** [review]
Keeping the hold in case we'd like to:
~~1. Fix one outstanding UX issue mentioned in https://github.com/gitpod-io/gitpod/pull/11023#discussion_r912011228.~~
2. Ask for another pair of eyes to take a closer look at the code changes. Cc @geropl

File: components/dashboard/src/settings/ProfileInformation.tsx
Comment (gtsiolis) [reply components/dashboard/src/settings/ProfileInformation.tsx:162]: Fair point—Let's keep it as is. 🤝

File: components/gitpod-protocol/src/protocol.ts
Comment (geropl) [inline components/gitpod-protocol/src/protocol.ts:163]: idea: Would it make sense to group these two attributes (companyName and emailAddress) under e.g. `profile`? The motivation would be clarity: `additionalUserData` is already pretty crowded, and is degenerating into a "kitchen sink" rather quickly...

File: components/gitpod-protocol/src/protocol.ts
Comment (geropl) [inline components/gitpod-protocol/src/protocol.ts:78]: While I like introducing capability of letting users change their email, and sharpening our concept for "primary email" (which is ill-conceived atm) this introduces room for fraud.
We rely upon these email-addresses to be validated (e.g. exist and are under the user's control) for student plans/bonusses.

If we merge this as is we'd have to prioritize separating these concepts.

File: components/dashboard/src/workspaces/Workspaces.tsx
Comment (geropl) [inline components/dashboard/src/workspaces/Workspaces.tsx:75]: nit: some comment that this is ok because the modal hides itself would be nice.

File: components/gitpod-protocol/src/protocol.ts
Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:78]: Maybe we could have a field `profile.email` that is chosen to be filled by the user, but only with one of the (validated) addresses from one of their auth providers.

Again, if we're fine with the temporary gap here, that can be a gap. We just need to prioritize it, then. :+1:

**Comment by geropl** [comment]
@svenefftinge From the issues I don't understand how urgent this is. I feel we have a good opportunity to nicely solve our handling of email addresses if we delay it until end of next week and ask @AlexTugarev to clean it up beforehand (sth he's keen on for some time already).
If the [temporary gap is fine](https://github.com/gitpod-io/gitpod/pull/11023#discussion_r912906186) we can merge as-is, though, and I will ask him to do it anyway, then.

File: components/gitpod-protocol/src/protocol.ts
Comment (svenefftinge) [reply components/gitpod-protocol/src/protocol.ts:78]: This function is not used for the student eligability, so I don't think we have that problem.

Comment (svenefftinge) [reply components/gitpod-protocol/src/protocol.ts:78]: See https://github.com/gitpod-io/gitpod/blob/7c0ecdbf86a3554a59770fc61d75f0a48f0336b0/components/server/ee/src/user/eligibility-service.ts#L69

**Comment by svenefftinge** [comment]
@geropl I have addressed the two nits and left a comment on the other feedback. I'd like to land this as it unblocks @phimae's work on the lead identification.

File: components/gitpod-protocol/src/protocol.ts
Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts]: Indeed, my bad! :+1:

**Review by geropl — APPROVED** [review]
Code LGTM, tested and works :+1:

**Pull request merged**
